api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+	//
+	// +optional
+	Service *ExternalDNSServiceSourceOptions `json:"service,omitempty"`
+
+	// CRD describes source configuration options specific
+	// to the CRD source resource.
+	//
+	// +optional
+	CRD *ExternalDNSCRDSourceOptions `json:"crd,omitempty"`
+
+	// OpenShiftRoute source currently has no unique configuration options.
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=Ignore;Allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the ExternalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// PublishPolicy determines what kinds of Service
+	// resources are published to the given DNS provider.
+	//
+	// The following publishing policies are available:
+	//
+	//  "PublishExternal": Only publish DNS records for
+	//  service that are externally reachable,
+	//  such as NodePort, ExternalName, and LoadBalancer
+	//  services. This is the default behavior of ExternalDNS.
+	//
+	//  "PublishInternalExternal": Publish DNS records
+	//   for externally reachable services, in addition to
+	//   internally reachable services, such as ClusterIP
+	//   services.
+	//
+	// +optional
+	PublishPolicy *ServicePublishPolicy `json:"publishPolicy,omitempty"`
+
+	// ServiceType determines what types of Service resources
+	// are watched by ExternalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP" (requires a publishPolicy of "PublishInternalExternal")
+	//
+	// One or more Service types can be specified, if desired.
+	//
+	// If no service types are provided, ExternalDNS will be
+	// configured to create DNS records for LoadBalancer services
+	// only by default.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={"LoadBalancer"}
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	ServiceType []corev1.ServiceType `json:"serviceType,omitempty"`
+}
+
+type ServicePublishPolicy string
+
+// +kubebuilder:validation:Enum=PublishExternal;PublishInternalExternal
+const (
+	ServicePublishPolicyExternalOnly        ServicePublishPolicy = "PublishExternal"
+	ServicePublishPolicyExternalAndInternal ServicePublishPolicy = "PublishInternalExternal"
+)
+
+type ExternalDNSCRDSourceOptions struct {
+	// Kind is the kind of the CRD
+	// source resource type to be
+	// consumed by ExternalDNS.
+	//
+	// e.g. "DNSEndpoint"
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Kind string `json:"kind"`
```

Comment:
Why use a kind instead of a resource?  Kinds are ambiguous mappings.  Resources are unambiguous.

Reply:
See https://github.com/openshift/external-dns-operator/pull/12#discussion_r665715937

Reply:
ExternalDNS lists all CRDs in the specified API group/version and uses the first one with a matching kind: https://github.com/kubernetes-sigs/external-dns/blob/2605c5ee93de4c04c62482211e887e7c7c3e95f5/source/crd.go#L82.  If we used resource here, the operator would have to translate resource to kind for ExternalDNS.  Using resource is semantically what we want, but using kind is what ExternalDNS does, so using resource here would add extra complexity and opportunities for failure.